File: Week_04/cpp/src/_874_WalkingRobotSimulation.cpp
#include <unordered_set>
#include <vector>
using namespace std;
struct pair_hash
{
    template <class T1, class T2>
    std::size_t operator()(const std::pair<T1, T2> &p) const
    {
        auto h1 = std::hash<T1>{}(p.first);
        auto h2 = std::hash<T2>{}(p.second);
        return h1 ^ h2;
    }
};

class Solution
{
public:
    // 注意使用unordered_set与pair需要自定义哈希函数
    // 方向的切换颇为巧妙
    int robotSim(vector<int> &commands, vector<vector<int>> &obstacles)
    {
        vector<pair<int, int>> dirs = {{0, 1}, {-1, 0}, {0, -1}, {1, 0}};
        unordered_set<pair<int, int>, pair_hash> obSet;
        for (auto &point : obstacles)
        {
            obSet.emplace(point[0], point[1]);
        }
        int ret = 0, di = 0, x = 0, y = 0;
        for (int cmd : commands)
        {
            if (cmd == -2)
            {
                di = (di + 1) % 4;
            }
            else if (cmd == -1)
            {
                di = (di + 3) % 4;
            }
            else
            {
                for (int i = 0; i < cmd; i++)
                {
                    auto &[dx, dy] = dirs[di];
                    int p = x + dx;
                    int q = y + dy;
                    if (obSet.find(make_pair(p, q)) != obSet.end())
                        break;
                    x = p;
                    y = q;
                    ret = max(ret, p * p + q * q);
                }
            }
        }
        return ret;
    }
};

File: Week_04/kotlin/src/_874_WalkingRobotSimulation.kt
class _874_Simulation {
    // O(n) n表示机器人行走的步数与转身数之和 O(m) m表示障碍物的个数
    fun robotSim(commands: IntArray, obstacles: Array<IntArray>): Int {
        val dirs = arrayOf(intArrayOf(0, 1), intArrayOf(-1, 0), intArrayOf(0, -1), intArrayOf(1, 0))
        var ret = 0
        var di = 0
        var x = 0
        var y = 0
        val map = hashMapOf<Int, HashSet<Int>>()
        obstacles.forEach { (x, y) -> map.computeIfAbsent(x) { HashSet() }.add(y) }
        for (cmd in commands) {
            when (cmd) {
                -2 -> di = (di + 1) % 4
                -1 -> di = (di + 3) % 4
                else -> {
                    for (i in 0 until cmd) {
                        val p = x + dirs[di][0]
                        val q = y + dirs[di][1]
                        if (p in map && q in map[p]!!) break;
                        x = p
                        y = q
                        ret = maxOf(ret, x * x + y * y)
                    }
                }
            }
        }
        return ret
    }
}